File: 백준/Gold/11505. 구간 곱 구하기/구간 곱 구하기.java
import java.io.*;
import java.util.*;

public class Main {

    static int MODULO = 1_000_000_007;

    public static class SegmentTree {
        long[] tree;
        int n;

        SegmentTree(long[] arr) {
            n = arr.length;
            tree = new long[4*n];
            build(arr, 1, 0, n-1);
        }

        private void build(long[] arr, int node, int start, int end) {
            if (start == end) {
                tree[node] = arr[start] % MODULO;
                return;
            }

            int mid = (start + end) / 2;
            build(arr, node * 2, start, mid);
            build(arr, node * 2 + 1, mid + 1, end);

            tree[node] = (tree[2*node] * tree[2*node+1]) % MODULO;
        }

        public void update(int idx, long val) {
            updateHelper(1, 0, n-1, idx, val);
        }

        public void updateHelper(int node, int start, int end, int idx, long val) {
            if (start == end) {
                tree[node] = val % MODULO;
                return;
            }

            int mid = (start + end) / 2;

            if (idx <= mid) {
                updateHelper(node * 2, start, mid, idx, val);
            }
            if (idx > mid) {
                updateHelper(node * 2 + 1, mid + 1, end, idx, val);
            }

            tree[node] = (tree[2 * node] * tree[2 * node + 1]) % MODULO;
        }


        public long query(int l, int r) {
            return queryHelper(1, 0, n-1, l, r);
        }
        private long queryHelper(int node, int start, int end, int l, int r) {
            //1. 구간을 벗어남
            if (start > r || end < l) {
                return 1;
            }

            // 2. 완전히 겹침
            if (start >= l && end <= r) {
                return tree[node];
            }

            // 3. 그 외
            int mid = (start + end) / 2;
            long left = queryHelper(2*node, start, mid, l, r);
            long right = queryHelper(2*node + 1, mid +1, end, l, r);

            return left * right % MODULO;
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());

        long[] arr = new long[n];
        for (int i = 0; i < n; i++) {
            arr[i] = Long.parseLong(br.readLine());
        }

        SegmentTree segTree = new SegmentTree(arr);

        for (int i = 0; i < m+k; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            long c = Long.parseLong(st.nextToken());

            // update
            if (a == 1) {
                segTree.update(b-1, c);
            }
            // query
            else if (a == 2) {
                sb.append(segTree.query(b-1, (int) c-1)).append("\n");
            }

        }

        System.out.print(sb);
    }
}

File: 백준/Gold/11505. 구간 곱 구하기/구간 곱 구하기.py
import sys

# 빠른 입력을 위한 설정
input = sys.stdin.read

def solve():
    # 전체 데이터를 한 번에 읽어와 토큰화
    data = input().split()
    if not data:
        return
    
    ptr = 0
    n = int(data[ptr]); ptr += 1
    m = int(data[ptr]); ptr += 1
    k = int(data[ptr]); ptr += 1
    
    MODULO = 1_000_000_007

    arr = []
    for _ in range(n):
        arr.append(int(data[ptr]))
        ptr += 1

    # 세그먼트 트리 배열 크기 설정 (4 * n)
    tree = [0] * (4 * n)

    # 1. 트리 구축 (Build)
    def build(node, start, end):
        if start == end:
            tree[node] = arr[start] % MODULO
            return
        
        mid = (start + end) // 2
        build(node * 2, start, mid)
        build(node * 2 + 1, mid + 1, end)
        tree[node] = (tree[node * 2] * tree[node * 2 + 1]) % MODULO

    # 2. 값 수정 (Update)
    def update(node, start, end, idx, val):
        if start == end:
            tree[node] = val % MODULO
            return
        
        mid = (start + end) // 2
        if idx <= mid:
            update(node * 2, start, mid, idx, val)
        else:
            update(node * 2 + 1, mid + 1, end, idx, val)
        
        tree[node] = (tree[node * 2] * tree[node * 2 + 1]) % MODULO

    # 3. 구간 곱 계산 (Query)
    def query(node, start, end, l, r):
        # 구간을 벗어난 경우 (곱셈이므로 1 반환)
        if start > r or end < l:
            return 1
        
        # 구간에 완전히 포함된 경우
        if l <= start and end <= r:
            return tree[node]
        
        mid = (start + end) // 2
        left_val = query(node * 2, start, mid, l, r)
        right_val = query(node * 2 + 1, mid + 1, end, l, r)
        
        return (left_val * right_val) % MODULO

    # 트리 초기화
    build(1, 0, n - 1)

    results = []
    # M + K 번의 명령 처리
    for _ in range(m + k):
        a = int(data[ptr]); ptr += 1
        b = int(data[ptr]); ptr += 1
        c = int(data[ptr]); ptr += 1
        
        if a == 1:
            # b번째 수를 c로 바꿈 (인덱스는 0부터 시작하므로 b-1)
            update(1, 0, n - 1, b - 1, c)
        elif a == 2:
            # b부터 c까지의 곱 (인덱스 b-1부터 c-1까지)
            results.append(str(query(1, 0, n - 1, b - 1, c - 1)))

    # 결과 한꺼번에 출력
    sys.stdout.write("\n".join(results) + "\n")

if __name__ == "__main__":
    # 파이썬의 재귀 깊이 제한 해제
    sys.setrecursionlimit(200000)
    solve()